src/core/shape_inference/include/reshape_shape_inference.hpp

```diff
@@ -0,0 +1,363 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+#pragma once
+
+#include "compare.hpp"
+#include "dimension_util.hpp"
+#include "openvino/core/dimension_tracker.hpp"
+#include "openvino/op/reshape.hpp"
+#include "utils.hpp"
+
+namespace ov {
+namespace op {
+namespace reshape {
+template <class T, class U = void>
+struct Product {};
+
+/** \brief Helper to resolve the input and output product for static dimensions. */
+template <class T>
+struct Product<T, typename std::enable_if<!std::is_same<T, Dimension>::value>::type> {
+    T in{1};
+    T out{1};
+
+    void update_in(const T& in_dim) {
+        in *= in_dim;
+    }
+
+    void update_out(const T& out_dim) {
+        out *= out_dim;
+    }
+
+    void set_inf() {
+        in = T(-1);
+        out = T(-1);
+    }
+
+    const T& get_static_in() const {
+        return in;
+    }
+
+    const T& get_static_out() const {
+        return out;
+    }
+
+    void calculate() {}
+};
+
+/** \brief Helper to resolve the input and output product for ov::Dimension (dynamic) dimensions. */
+template <class T>
+struct Product<T, typename std::enable_if<std::is_same<T, Dimension>::value>::type> {
+    std::pair<T, T> in{1, 1};
+    std::pair<T, T> out{1, 1};
+
+    void update_in(const T& in_dim) {
+        inputs.emplace_back(in_dim);
+    }
+
+    void update_out(const T& out_dim) {
+        outputs.emplace_back(out_dim);
+    }
+
+    void set_inf() {
+        in.second = T(-1);
+        out.second = T(-1);
+    }
+
+    const T& get_static_in() const {
+        return in.first;
+    }
+
+    const T& get_static_out() const {
+        return out.first;
+    }
+
+    const T& get_dynamic_in() const {
+        return in.second;
+    }
+
+    const T& get_dynamic_out() const {
+        return out.second;
+    }
+
+    void calculate() {
+        // dimensions compare to remove same from product calculation
+        auto dim_full_eq = [](const T& lhs, const T& rhs) -> bool {
+            return (lhs == rhs) && DimensionTracker::get_label(lhs) == DimensionTracker::get_label(rhs) &&
+                   (lhs.is_static() || DimensionTracker::has_label(lhs));
```

Comment:
two static dims may be equal but with different labels -- shouldn't we handle them as equal anyways?

Reply:
Agree that for product calculation the  static shape with different labels has can be skipped. 
The intention is use these static dimensions into product calculation because of label propagation. ( in some cases when resolve label for '-1' the label can be passed and not always lost).